File: ai_safety_engine.py
import sqlite3
import json
import math
from datetime import datetime, timedelta
from contextlib import contextmanager

DATABASE = 'data/tourist_tracking.db'

@contextmanager
def get_db():
    """Context manager for database connections"""
    conn = sqlite3.connect(DATABASE)
    conn.row_factory = sqlite3.Row
    try:
        yield conn
        conn.commit()
    except Exception as e:
        conn.rollback()
        raise e
    finally:
        conn.close()

def calculate_distance(lat1, lon1, lat2, lon2):
    """Calculate distance between two coordinates in meters using Haversine formula"""
    R = 6371000  # Earth radius in meters
    
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)
    
    a = math.sin(delta_phi/2)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    
    return R * c

def analyze_tourist_safety(tourist_id):
    """Analyze safety for a specific tourist using rule-based AI"""
    anomalies = []
    risk_score = 0
    
    try:
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Get recent location history (last 24 hours)
            cursor.execute('''
                SELECT latitude, longitude, speed, timestamp
                FROM locations
                WHERE tourist_id = ? 
                AND timestamp > datetime('now', '-24 hours')
                ORDER BY timestamp DESC
                LIMIT 100
            ''', (tourist_id,))
            
            locations = cursor.fetchall()
            
            if len(locations) < 2:
                # Not enough data to analyze
                return update_safety_score(tourist_id, 'low', 0, [])
            
            # Get active SOS alerts count
            cursor.execute('''
                SELECT COUNT(*) as sos_count
                FROM sos_alerts
                WHERE tourist_id = ? AND status = 'active'
            ''', (tourist_id,))
            
            sos_count = cursor.fetchone()['sos_count']
            
            # Analysis 1: Check for active SOS alerts
            if sos_count > 0:
                anomalies.append('Active SOS Alert')
                risk_score += 50
            
            # Analysis 2: Check for sudden stops (stationary for too long)
            recent_locations = locations[:10]  # Last 10 locations
            if len(recent_locations) >= 3:
                is_stationary = True
                first_loc = recent_locations[0]
                
                for loc in recent_locations[1:]:
                    distance = calculate_distance(
                        first_loc['latitude'], first_loc['longitude'],
                        loc['latitude'], loc['longitude']
                    )
                    if distance > 100:  # More than 100 meters movement
                        is_stationary = False
                        break
                
                if is_stationary:
                    # Check if stationary for more than 2 hours
                    time_diff = datetime.fromisoformat(recent_locations[0]['timestamp']) - \
                               datetime.fromisoformat(recent_locations[-1]['timestamp'])
                    
                    if time_diff.total_seconds() > 7200:  # 2 hours
                        anomalies.append('Stationary for extended period')
                        risk_score += 15
            
            # Analysis 3: Check for unusual speed patterns
            speeds = [loc['speed'] for loc in locations if loc['speed'] is not None]
            if speeds:
                avg_speed = sum(speeds) / len(speeds)
                max_speed = max(speeds)
                
                # Check for unusually high speed (over 150 km/h)
                if max_speed and max_speed > 41.67:  # 150 km/h in m/s
                    anomalies.append('Unusually high speed detected')
                    risk_score += 20
            
            # Analysis 4: Check for erratic movement (frequent direction changes)
            if len(locations) >= 5:
                direction_changes = 0
                for i in range(len(locations) - 2):
                    loc1 = locations[i]
                    loc2 = locations[i + 1]
                    loc3 = locations[i + 2]
                    
                    # Calculate bearing between points
                    angle1 = calculate_bearing(
                        loc1['latitude'], loc1['longitude'],
                        loc2['latitude'], loc2['longitude']
                    )
                    angle2 = calculate_bearing(
                        loc2['latitude'], loc2['longitude'],
                        loc3['latitude'], loc3['longitude']
                    )
                    
                    # Check for significant direction change (>90 degrees)
                    angle_diff = abs(angle1 - angle2)
                    if angle_diff > 90 and angle_diff < 270:
                        direction_changes += 1
                
                if direction_changes >= 3:
                    anomalies.append('Erratic movement pattern')
                    risk_score += 10
            
            # Analysis 5: Check for route deviation (moved far from last known area)
            if len(locations) >= 20:
                recent_area = locations[:5]
                older_area = locations[-10:]
                
                # Calculate centroid of recent area
                recent_lat = sum([loc['latitude'] for loc in recent_area]) / len(recent_area)
                recent_lon = sum([loc['longitude'] for loc in recent_area]) / len(recent_area)
                
                # Calculate centroid of older area
                older_lat = sum([loc['latitude'] for loc in older_area]) / len(older_area)
                older_lon = sum([loc['longitude'] for loc in older_area]) / len(older_area)
                
                # Check distance between centroids
                deviation_distance = calculate_distance(recent_lat, recent_lon, older_lat, older_lon)
                
                if deviation_distance > 50000:  # More than 50km deviation
                    anomalies.append('Significant route deviation')
                    risk_score += 15
            
            # Analysis 6: Check for no recent updates (lost contact)
            latest_location = locations[0]
            time_since_update = datetime.now() - datetime.fromisoformat(latest_location['timestamp'])
            
            if time_since_update.total_seconds() > 3600:  # More than 1 hour
                anomalies.append('No recent location updates')
                risk_score += 25
            elif time_since_update.total_seconds() > 1800:  # More than 30 minutes
                anomalies.append('Delayed location updates')
                risk_score += 10
            
            # Determine risk level based on score
            if risk_score >= 50:
                risk_level = 'critical'
            elif risk_score >= 30:
                risk_level = 'high'
            elif risk_score >= 15:
                risk_level = 'medium'
            else:
                risk_level = 'low'
            
            # Update safety score in database
            return update_safety_score(tourist_id, risk_level, risk_score, anomalies)
    
    except Exception as e:
        print(f"[v0] AI Analysis error for tourist {tourist_id}: {str(e)}")
        return False

def calculate_bearing(lat1, lon1, lat2, lon2):
    """Calculate bearing between two coordinates in degrees"""
    phi1 = math.radians(lat1)
    phi2 = math.radians(lat2)
    delta_lambda = math.radians(lon2 - lon1)
    
    y = math.sin(delta_lambda) * math.cos(phi2)
    x = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(delta_lambda)
    
    bearing = math.degrees(math.atan2(y, x))
    return (bearing + 360) % 360

def update_safety_score(tourist_id, risk_level, risk_score, anomalies):
    """Update safety score in database"""
    try:
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Check if safety score exists
            cursor.execute('SELECT id FROM safety_scores WHERE tourist_id = ?', (tourist_id,))
            existing = cursor.fetchone()
            
            anomalies_json = json.dumps(anomalies)
            
            if existing:
                cursor.execute('''
                    UPDATE safety_scores 
                    SET risk_level = ?, risk_score = ?, anomalies = ?, last_calculated = CURRENT_TIMESTAMP
                    WHERE tourist_id = ?
                ''', (risk_level, risk_score, anomalies_json, tourist_id))
            else:
                cursor.execute('''
                    INSERT INTO safety_scores (tourist_id, risk_level, risk_score, anomalies)
                    VALUES (?, ?, ?, ?)
                ''', (tourist_id, risk_level, risk_score, anomalies_json))
            
            return True
    
    except Exception as e:
        print(f"[v0] Update safety score error: {str(e)}")
        return False

def analyze_all_tourists():
    """Analyze safety for all active tourists"""
    try:
        with get_db() as conn:
            cursor = conn.cursor()
            cursor.execute('SELECT id FROM tourists WHERE is_active = 1')
            tourists = cursor.fetchall()
            
            for tourist in tourists:
                analyze_tourist_safety(tourist['id'])
                print(f"[v0] Analyzed tourist {tourist['id']}")
            
            print(f"[v0] Analyzed {len(tourists)} tourists")
            return True
    
    except Exception as e:
        print(f"[v0] Analyze all tourists error: {str(e)}")
        return False

if __name__ == '__main__':
    print("[v0] Starting AI Safety Analysis...")
    analyze_all_tourists()
    print("[v0] Analysis complete")

File: lib/ai-safety-engine.ts
import { getDb, type Location } from "./db"

interface MovementAnalysis {
  suddenStop: boolean
  unusualSpeed: boolean
  routeDeviation: boolean
  frequentSOS: boolean
  stationaryTooLong: boolean
}

interface RiskAssessment {
  risk_level: "low" | "medium" | "high" | "critical"
  risk_score: number
  anomaly_type?: string
  details?: string
}

export class AISafetyEngine {
  // Thresholds for anomaly detection
  private readonly SPEED_THRESHOLD_KMH = 150 // Unrealistic speed
  private readonly STATIONARY_THRESHOLD_MINUTES = 120 // 2 hours
  private readonly DEVIATION_THRESHOLD_METERS = 5000 // 5km from expected route
  private readonly FREQUENT_SOS_COUNT = 2 // Multiple SOS in short time
  private readonly SUDDEN_STOP_SPEED_DROP = 0.9 // 90% speed drop

  /**
   * Calculate distance between two coordinates using Haversine formula
   */
  private calculateDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
    const R = 6371e3 // Earth radius in meters
    const φ1 = (lat1 * Math.PI) / 180
    const φ2 = (lat2 * Math.PI) / 180
    const Δφ = ((lat2 - lat1) * Math.PI) / 180
    const Δλ = ((lon2 - lon1) * Math.PI) / 180

    const a = Math.sin(Δφ / 2) * Math.sin(Δφ / 2) + Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2)

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))

    return R * c // Distance in meters
  }

  /**
   * Calculate speed between two location points
   */
  private calculateSpeed(loc1: Location, loc2: Location): number {
    const distance = this.calculateDistance(loc1.latitude, loc1.longitude, loc2.latitude, loc2.longitude)

    const time1 = new Date(loc1.timestamp).getTime()
    const time2 = new Date(loc2.timestamp).getTime()
    const timeDiffHours = Math.abs(time2 - time1) / (1000 * 60 * 60)

    if (timeDiffHours === 0) return 0

    return distance / 1000 / timeDiffHours // km/h
  }

  /**
   * Analyze movement patterns for a specific tourist
   */
  async analyzeMovement(touristId: number): Promise<MovementAnalysis> {
    const db = await getDb()

    // Get recent locations (last 24 hours)
    const locations = await db.all<Location[]>(
      `SELECT * FROM locations 
       WHERE tourist_id = ? 
       AND timestamp > datetime('now', '-24 hours')
       ORDER BY timestamp DESC 
       LIMIT 50`,
      touristId,
    )

    if (locations.length < 2) {
      return {
        suddenStop: false,
        unusualSpeed: false,
        routeDeviation: false,
        frequentSOS: false,
        stationaryTooLong: false,
      }
    }

    // Check for sudden stop
    const suddenStop = this.detectSuddenStop(locations)

    // Check for unusual speed
    const unusualSpeed = this.detectUnusualSpeed(locations)

    // Check for stationary too long
    const stationaryTooLong = this.detectStationaryTooLong(locations)

    // Check for frequent SOS
    const frequentSOS = await this.detectFrequentSOS(touristId)

    // Check for route deviation (simplified - would need expected routes in production)
    const routeDeviation = this.detectRouteDeviation(locations)

    return {
      suddenStop,
      unusualSpeed,
      routeDeviation,
      frequentSOS,
      stationaryTooLong,
    }
  }

  /**
   * Detect if tourist suddenly stopped moving
   */
  private detectSuddenStop(locations: Location[]): boolean {
    if (locations.length < 3) return false

    const recent = locations.slice(0, 3)
    const speed1 = this.calculateSpeed(recent[2], recent[1])
    const speed2 = this.calculateSpeed(recent[1], recent[0])

    // If speed drops by more than 90% suddenly
    if (speed1 > 10 && speed2 < speed1 * (1 - this.SUDDEN_STOP_SPEED_DROP)) {
      return true
    }

    return false
  }

  /**
   * Detect unrealistic or dangerous speeds
   */
  private detectUnusualSpeed(locations: Location[]): boolean {
    for (let i = 0; i < locations.length - 1; i++) {
      const speed = this.calculateSpeed(locations[i + 1], locations[i])
      if (speed > this.SPEED_THRESHOLD_KMH) {
        return true
      }
    }
    return false
  }

  /**
   * Detect if tourist has been stationary for too long
   */
  private detectStationaryTooLong(locations: Location[]): boolean {
    if (locations.length < 5) return false

    const recent = locations.slice(0, 5)
    const firstLoc = recent[recent.length - 1]
    const lastLoc = recent[0]

    const distance = this.calculateDistance(firstLoc.latitude, firstLoc.longitude, lastLoc.latitude, lastLoc.longitude)

    const timeDiff = (new Date(lastLoc.timestamp).getTime() - new Date(firstLoc.timestamp).getTime()) / (1000 * 60) // minutes

    // If moved less than 100m in 2+ hours
    if (distance < 100 && timeDiff > this.STATIONARY_THRESHOLD_MINUTES) {
      return true
    }

    return false
  }

  /**
   * Detect frequent SOS alerts
   */
  private async detectFrequentSOS(touristId: number): Promise<boolean> {
    const db = await getDb()

    const sosCount = await db.get(
      `SELECT COUNT(*) as count 
       FROM sos_alerts 
       WHERE tourist_id = ? 
       AND timestamp > datetime('now', '-6 hours')`,
      touristId,
    )

    return (sosCount?.count || 0) >= this.FREQUENT_SOS_COUNT
  }

  /**
   * Detect significant route deviation
   */
  private detectRouteDeviation(locations: Location[]): boolean {
    if (locations.length < 10) return false

    // Calculate average position
    const avgLat = locations.reduce((sum, loc) => sum + loc.latitude, 0) / locations.length
    const avgLon = locations.reduce((sum, loc) => sum + loc.longitude, 0) / locations.length

    // Check if any recent location is too far from average
    const recentLocs = locations.slice(0, 3)
    for (const loc of recentLocs) {
      const distance = this.calculateDistance(avgLat, avgLon, loc.latitude, loc.longitude)
      if (distance > this.DEVIATION_THRESHOLD_METERS) {
        return true
      }
    }

    return false
  }

  /**
   * Calculate overall risk score and level
   */
  calculateRiskScore(analysis: MovementAnalysis): RiskAssessment {
    let score = 0
    const anomalies: string[] = []

    if (analysis.frequentSOS) {
      score += 40
      anomalies.push("Frequent SOS alerts")
    }

    if (analysis.suddenStop) {
      score += 25
      anomalies.push("Sudden stop detected")
    }

    if (analysis.unusualSpeed) {
      score += 20
      anomalies.push("Unusual speed pattern")
    }

    if (analysis.stationaryTooLong) {
      score += 30
      anomalies.push("Stationary too long")
    }

    if (analysis.routeDeviation) {
      score += 15
      anomalies.push("Route deviation")
    }

    let risk_level: "low" | "medium" | "high" | "critical"
    if (score >= 60) {
      risk_level = "critical"
    } else if (score >= 40) {
      risk_level = "high"
    } else if (score >= 20) {
      risk_level = "medium"
    } else {
      risk_level = "low"
    }

    return {
      risk_level,
      risk_score: score,
      anomaly_type: anomalies.length > 0 ? anomalies[0] : undefined,
      details: anomalies.join(", ") || "No anomalies detected",
    }
  }

  /**
   * Process safety analysis for a tourist and store results
   */
  async processTouristSafety(touristId: number): Promise<void> {
    const analysis = await this.analyzeMovement(touristId)
    const riskAssessment = this.calculateRiskScore(analysis)

    const db = await getDb()

    await db.run(
      `INSERT INTO safety_scores (tourist_id, risk_level, risk_score, anomaly_type, details, timestamp)
       VALUES (?, ?, ?, ?, ?, datetime('now'))`,
      [
        touristId,
        riskAssessment.risk_level,
        riskAssessment.risk_score,
        riskAssessment.anomaly_type,
        riskAssessment.details,
      ],
    )

    console.log(`[v0] Safety analysis for tourist ${touristId}:`, riskAssessment)
  }

  /**
   * Process all active tourists
   */
  async processAllTourists(): Promise<void> {
    const db = await getDb()

    const tourists = await db.all(`
      SELECT DISTINCT t.id 
      FROM tourists t
      JOIN locations l ON t.id = l.tourist_id
      WHERE l.timestamp > datetime('now', '-24 hours')
    `)

    for (const tourist of tourists) {
      await this.processTouristSafety(tourist.id)
    }

    console.log(`[v0] Processed safety analysis for ${tourists.length} tourists`)
  }
}
